projects/packages/post-list/src/class-admin.php

```diff
@@ -179,22 +179,40 @@ public function enqueue_scripts( $hook ) {
 	 * @param int $post_id The current post ID.
 	 * @return array The featured image id and URLs
 	 */
-	protected static function get_featured_or_first_post_image( $post ) {
-		$post_id = $post->ID;
-
+	protected function get_featured_or_first_post_image( $post ) {
 		$featured_image_id    = null;
 		$featured_image_url   = null;
 		$featured_image_thumb = null;
 
+		$post_id = $post->ID;
+
+		// If a featured image exists for the post, use that thumbnail.
 		if ( has_post_thumbnail() ) {
 			$featured_image_id    = get_post_thumbnail_id( $post_id );
 			$featured_image_url   = get_the_post_thumbnail_url( $post_id );
 			$featured_image_thumb = get_the_post_thumbnail_url( $post_id, array( 50, 50 ) );
 		} else {
-			$image = current( self::get_content_images( $post->post_content ) );
-			if ( ! empty( $image ) ) {
-				$featured_image_url   = $image['uri'];
-				$featured_image_thumb = $image['uri'];
+			// If a featured image does not exist look for the first "media library" hosted image on the post.
+			$dom = new \DOMDocument();
+
+			// libxml_use_internal_errors(true) silences PHP warnings and errors from malformed HTML in loadHTML().
+			// you can consult libxml_get_last_error() or libxml_get_errors() to check for errors if needed.
+			libxml_use_internal_errors( true );
+			$dom->loadHTML( $post->post_content );
+
+			// Media library images have a class attribute value containing 'wp-image-{$attachment_id}'.
+			// Use DomXPath to parse the post content and get the first img tag containing 'wp-image-' as a class value.
+			$class_name = 'wp-image-';
+			$dom_x_path = new \DomXPath( $dom );
+			$nodes      = $dom_x_path->query( "//img[contains(@class, '$class_name')]/@class" );
+
+			if ( $nodes->length > 0 ) {
+				// Get the $attachment_id from the end of the class name value of the 1st image node (aka index 0).
+				$attachment_id = str_replace( $class_name, '', $nodes[0]->value );
+
+				$featured_image_id    = $attachment_id;
+				$featured_image_url   = wp_get_attachment_image_url( $attachment_id, 'full-size' );
+				$featured_image_thumb = wp_get_attachment_image_url( $attachment_id, array( 50, 50 ) );
```

Comment:
Instead of using regular expression to parse the HTML, I opted to use DOMDocument and DomXPath. It's a little more code, but I think it's a more robust solution. For example DomXPath won't match on a random string inside a paragraph tag, but a regular expression probably would.

In either case, this solution hinges on 'wp-image-' being inside the img class attribute. If that value changes in the future this code will return 0 matches. Given the lack of alternatives, I believe this is still a worthwhile approach. We might can hedge against this with a unit test that confirms media library images contain 'wp-image-' when added to a post. Additionally, worse case scenario, the featured image thumbnail is our main condition, and it is un-affected by this code.

Reply:
> wp-image-

Just double checking - will this work for images inserted both in image blocks, but also within Classic Blocks ,or the Classic Editor?

Reply:
yes, it will work as you described. The function defined there inspects HTML text patterns that defined the existence of images in the post content code. If found, will pick the first one which will be used to represent the thumbnail in the post list page.

It isn't perfect and should be taken mostly like a fallback of the post featured-image when it isn't defined. Finally, it's pretty similar o what Calypso does. In theory, it shouldn't be worse :-)

Reply:
@DustyReagan looks cleaner and reliable, and also with the ability to get the post thumbnail. 👏 
I wonder whether in case no images with the class name shape as defined in the code,  aren't found, can we add a third fallback just inspecting images?

We should try to avoid DOM-friendly processes as much as possible. Pretty sure you're already aware and it's a limitation that we'll have to live with, but a third option would be nice too. You could use the same DOMXPath approach surely.

Reply:
> looks cleaner and reliable, and also with the ability to get the post thumbnail. 👏

Thanks!

> I wonder whether in case no images with the class name shape as defined in the code, aren't found, can we add a third fallback just inspecting images?

Do you mean as a 3rd option, display an image hosted outside of media library if one exists? We could do that, but we won't have thumbnail sized resolution version of the image. We could end up with a massive image scaled down to thumbnail size via CSS in the post-list. We could produce a thumbnail and store it, or maybe add the image to media library automatically and reference the thumbnail it creates, or maybe even just create the thumbnail on the fly in real-time. It's possible! But is it too big of an undertaking?

Reply:
Let me mark this comment as resolved for now. This is definitely a nice idea but something to implement, in case we want, in a follow-up task, right?